src/mongo/base/CompositeBase.class.php

```diff
@@ -0,0 +1,164 @@
+            if(!empty($spec)){
+                $docHash = md5($doc[_ID_KEY][_ID_RESOURCE] . $doc[_ID_KEY][_ID_CONTEXT]);
+
+                if(!array_key_exists($docHash, $candidates)){
+                    $candidates[$docHash] = array(
+                        'id'=>array(
+                            _ID_RESOURCE=>$doc[_ID_KEY][_ID_RESOURCE],
+                            _ID_CONTEXT=>$doc[_ID_KEY][_ID_CONTEXT],
+                        )
+                    );
+                }
+                if(!array_key_exists('specTypes', $candidates[$docHash])) {
+                    $candidates[$docHash]['specTypes'] = array();
+                }
+                // Save the specification type so we only have to regen resources in that table type
+                if(!in_array($doc[_ID_KEY][_ID_TYPE], $candidates[$docHash]['specTypes']))
+                {
+                    $candidates[$docHash]['specTypes'][] = $doc[_ID_KEY][_ID_TYPE];
+                }
+            }
+        }
+
+        // convert operations to subjects
+        $impactedSubjects = array();
+        foreach($candidates as $candidate){
+            if(in_array($candidate['id'][_ID_RESOURCE], $deletedSubjects)){
+                $candidate['delete'] = true;
+            } else {
+                $candidate['delete'] = false;
+            }
+            $specTypes = (isset($candidate['specTypes']) ? $candidate['specTypes'] : array());
+            $impactedSubjects[] = new ImpactedSubject($candidate['id'], $this->getOperationType(), $this->getStoreName(), $this->getPodName(), $specTypes, $candidate['delete']);
+        }
+        return $impactedSubjects;
+    }
+
+    public abstract function getTypesInSpecification();
+
+    public abstract function findImpactedComposites($resourcesAndPredicates,$contextAlias);
+
+    public abstract function getSpecification($storeName, $composite_id);
+
+    /**
+     * Test if the a particular type appears in the array of types associated with a particular spec and that the changeset
+     * includes rdf:type (or is empty, meaning addition or deletion vs. update)
+     * @param string $rdfType
+     * @param array $validTypes
+     * @param array|null $subjectPredicates
+     * @return bool
+     */
+    protected function checkIfTypeShouldTriggerOperation($rdfType, array $validTypes, $subjectPredicates)
+    {
+        //todo: views wasn't using this code. Figure out why.
+        // We don't know if this is an alias or a fqURI, nor what is in the valid types, necessarily
+        $types = array($rdfType);
+        try
+        {
+            $types[] = $this->labeller->qname_to_uri($rdfType);
+        }
+        catch(TripodLabellerException $e) {}
+        try
+        {
+            $types[] = $this->labeller->uri_to_alias($rdfType);
+        }
+        catch(TripodLabellerException $e) {}
+
+        $intersectingTypes = array_unique(array_intersect($types, $validTypes));
+        if(!empty($intersectingTypes))
```

Comment:
I don't think this block is what we want, but I'm not 100% sure.  It *looks* like it will only return true here if:
1. The document's rdf:type intersects with the specification's
2. AND
  * The array of predicates is empty *OR*
  * The array of predicates contains rdf:type as either an alias or FQ URI

I *think* we always want to return true if there's an intersection, else false.

Can you think of any examples where that isn't true?